File: 卡码/代码随想录/C++/图论/graph/dfs/LeetCode_0200.cpp
//200. 岛屿数量
#include <iostream>
#include <vector>
using namespace std;

class Solution {
	int dir[4][2]={{0,1},{1,0},{0,-1},{-1,0}};
public:
	void dfs(vector<vector<char>>& grid,int x,int y){
		grid[x][y]='0';
		for(int i=0;i<4;i++){
			int nextX=x+dir[i][0];
			int nextY=y+dir[i][1];
			if(nextX<0||nextX>=grid.size()||nextY<0||nextY>=grid[0].size()) continue;
			if(grid[nextX][nextY]=='1'){
				dfs(grid,nextX,nextY);
			}
		}
	}
	int numIslands(vector<vector<char>>& grid) {
		int n=grid.size(),m=grid[0].size(),res=0;
		for(int i=0;i<n;i++){
			for(int j=0;j<m;j++){
				if(grid[i][j]=='1'){
					res++;
					dfs(grid,i,j);
				}
			}
		}
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/图论/dfs/LeetCode_0200.java
package com.kurumi.leetCode.codeThoughts.图论.dfs;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 200. 岛屿数量
 */
public class LeetCode_0200 {
    static class Solution {
        int[][] dir = new int[][] { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };

        public void dfs(char[][] grid, int x, int y) {
            grid[x][y] = '0';
            for (int i = 0; i < 4; i++) {
                int nextX = x + dir[i][0];
                int nextY = y + dir[i][1];
                if (nextX < 0 || nextX >= grid.length || nextY < 0 || nextY >= grid[0].length)
                    continue;
                if (grid[nextX][nextY] == '1') {
                    grid[nextX][nextY] = '0';
                    dfs(grid, nextX, nextY);
                }
            }
        }

        public int numIslands(char[][] grid) {
            int n = grid.length;
            int m = grid[0].length;
            int res = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (grid[i][j] == '1') {
                        res++;
                        dfs(grid, i, j);
                    }
                }
            }
            return res;
        }
    }
}
